src/components/ActivityCollector/attachClickActivityCollector.js

```diff
@@ -12,16 +12,34 @@ governing permissions and limitations under the License.
 
 import { noop } from "../../utils";
 
-const createClickHandler = ({ eventManager, lifecycle, handleError }) => {
+const createClickHandler = ({
+  eventManager,
+  lifecycle,
+  handleError,
+  onBeforeLinkCollect
+}) => {
   return clickEvent => {
     // TODO: Consider safeguarding from the same object being clicked multiple times in rapid succession?
     const clickedElement = clickEvent.target;
     const event = eventManager.createEvent();
+
     return (
       lifecycle
         .onClick({ event, clickedElement })
         .then(() => {
-          if (event.isEmpty()) {
+          let shouldCollectLinkClick = null;
+          if (onBeforeLinkCollect) {
+            shouldCollectLinkClick = onBeforeLinkCollect({
+              xdm: event.xdm ? event.xdm : null,
```

Comment:
From today's standup discussion, but maybe modifying our conclusion a bit, I think you would create a new method on event called `executeOnBeforeLinkClick`. From here in attachClickActivityCollector.js, you would then call it:

```
event.executeOnBeforeLinkCollect(onBeforeLinkCollect);
```

And event would execute `onBeforeLinkCollect`.

You might run into the question of which XDM do we pass into `onBeforeLinkCollect`. We have this concept of user-provided XDM vs automatically-collected XDM. What @jonsnyder was saying during standup was that the XDM that we generate in https://github.com/adobe/alloy/blob/e30caf3a6f520410221b6008f797561b641a23b1/src/components/ActivityCollector/createLinkClick.js#L60-L72 is really acting like "user-provided data", since we're the one generating the event on behalf of the customer. Usually user-provided XDM is added to the event using `event.setUserXdm`. 

Right now, createLinkClick.js is calling `event.mergeXdm` instead of `event.setUserXdm`. We could switch it to call `event.setUserXdm` and then when event calls `onBeforeLinkCollect`, it could provide `onBeforeLinkCollect` with the "user XDM", which is currently stored in a `userXdm` variable inside event. We'll need to make sure this doesn't cause other problems though. I would poke around the scenario where a user clicks a link that's also part of a Personalization offer, since Personalization also attempts to augment the event XDM (https://github.com/adobe/alloy/blob/e30caf3a6f520410221b6008f797561b641a23b1/src/components/Personalization/createOnClickHandler.js#L20-L21).

Anyway, go with that and see where you get. It's probably worth bringing up your findings in standup or a tech checkin and just validate that we're all on the same page.